src/session_impl.cpp

```diff
@@ -1694,7 +1695,8 @@ namespace aux {
 	void session_impl::on_ip_change(error_code const& ec)
 	{
 		if (ec) return;
-		m_ip_notifier.async_wait(std::bind(&session_impl::on_ip_change, this, _1));
+		m_ip_notifier.async_wait([this] (error_code const& ec)
+			{ this->wrap(&session_impl::on_ip_change, ec); });
 		reopen_listen_sockets();
```

Comment:
I suppose in this specific case, it's OK to invoke this handler "recursively", but in general I'm pretty sure the convention is to always post handlers.